### PR TITLE
[contrib] locust stress test environment setup based on ansible

### PR DESCRIPTION
**Pull request by ydye**

### CODE REVIEW

File: contrib/api-stress-test/readme.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      operator: "Equal",
+      value: "mock",
+      effect: "NoSchedule",
+    }
+]
+```
+
+2. Build image and push them into a registry 
+
+#### Write Inventory
```
[the remark this answers is not in the document]
Comment (ydye) [reply contrib/api-stress-test/readme.md:27]: OpenPai will be deployed by the playbooks. K8S is necessary

File: contrib/api-stress-test/roles/generate/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+default_vk_per_host: "{{ vk_per_host - stress_vk_per_host |int }}"
+
+defaulvc_numbers: "{{ stress_vk_per_host |int * vk_host_numbers |int }}"
+stressvc_numbers: "{{ default_vk_per_host |int * vk_host_numbers |int }}"
+
+pylon_domain: "{{ hostvars[groups['kube-master'][0]].ansible_host }}"
+
+docker_registry_namespace: openpai
+docker_registry_domain: docker.io
+docker_registry_tag: stresstest
```
[the remark this answers is not in the document]
Comment (ydye) [reply contrib/api-stress-test/roles/generate/defaults/main.yml:15]: Yes, I have push the image into openpai repo

File: contrib/api-stress-test/roles/locust/config/templates/stress.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    max_retries = {{ locust_http_max_retries }}
+    connection_timeout = {{ locust_http_connection_timeout }}
+    network_timeout = {{ locust_http_network_timeout }}
+
+    def on_start(self):
+        self.userid = str(uuid.uuid4())
+
+{% if stress_test_openpai_job_submit_enable %}
+    @task({{ test_api_list.openpai_job_submit.weight }})
+    def submitjob(self):
```
Comment (Binyang2014) [inline contrib/api-stress-test/roles/locust/config/templates/stress.py:66]: rename to `submit_job`?

File: contrib/api-stress-test/roles/locust/config/templates/stress.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        self.client.post(
+            "/rest-server/api/v2/jobs",
+            headers=openpai_headers,
+            data=template_data
+        )
+{% endif %}
+
+{% if stress_test_openpai_job_list_enable %}
+    @task({{ test_api_list.openpai_job_list.weight }})
+    def listjoball(self):
```
Comment (Binyang2014) [inline contrib/api-stress-test/roles/locust/config/templates/stress.py:86]: rename to `list_job_all`?

File: contrib/api-stress-test/roles/locust/config/templates/stress.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        self.client.get(
+            "/rest-server/api/v2/jobs",
+            headers=openpai_headers
+        )
+{% endif %}
+
+{% if stress_test_k8s_pod_list_enable %}
+    @task({{ test_api_list.k8s_pod_list.weight }})
+    def getPodList(self):
```
Comment (Binyang2014) [inline contrib/api-stress-test/roles/locust/config/templates/stress.py:98]: rename to `get_pod_list`?

File: contrib/api-stress-test/roles/openpai/install/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+docker_registry_namespace: openpai
+docker_registry_domain: docker.io
+docker_registry_tag: stresstest
+
+#docker_registry_username: example
+#docker_registry_password: example
+
+kube_config_path: "{{local_home_path}}/.kube/config"
+
+openpai_branch_name: "stress-test"
```
Comment (Binyang2014) [inline contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: Why not use `master` branch. Do we plan to maintain the branch?

File: contrib/api-stress-test/readme.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ydye) [reply contrib/api-stress-test/readme.md:43]: Done

File: contrib/api-stress-test/readme.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+1. Please create your own branch and modify rest-server's code for vk scheduling.
+
+- Remove the following line
+
+Line: https://github.com/microsoft/pai/blob/master/src/rest-server/src/models/v2/job/k8s.js#L465
+
+- add the toleration configuration after this line.
+
+Line: https://github.com/microsoft/pai/blob/master/src/rest-server/src/models/v2/job/k8s.js#L416
```
[the remark this answers is not in the document]
Comment (ydye) [reply contrib/api-stress-test/readme.md]: Done

File: contrib/api-stress-test/readme.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
+#### Before starting the stress test playbooks to setup environment.
+
+1. Please create your own branch and modify rest-server's code for vk scheduling.
+
+- Remove the following line
+
+Line: https://github.com/microsoft/pai/blob/master/src/rest-server/src/models/v2/job/k8s.js#L465
```
[the remark this answers is not in the document]
Comment (ydye) [reply contrib/api-stress-test/readme.md]: Done

File: contrib/api-stress-test/roles/openpai/install/defaults/main.yml (hunk 6 above)
Comment (ydye) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: No, but this branch has been modified for vk scheduling. And can be used directly.

File: contrib/api-stress-test/roles/locust/config/templates/stress.py (hunk 3 above)
Comment (ydye) [reply contrib/api-stress-test/roles/locust/config/templates/stress.py:66]: Done

File: contrib/api-stress-test/roles/locust/config/templates/stress.py (hunk 4 above)
Comment (ydye) [reply contrib/api-stress-test/roles/locust/config/templates/stress.py:86]: Done

File: contrib/api-stress-test/roles/locust/config/templates/stress.py (hunk 5 above)
Comment (ydye) [reply contrib/api-stress-test/roles/locust/config/templates/stress.py:98]: Done

File: contrib/api-stress-test/roles/locust/config/templates/stress.py (hunk 3 above)
Comment (Binyang2014) [reply contrib/api-stress-test/roles/locust/config/templates/stress.py:66]: ...I mean the python function name. rename `def submitjob` to `def submit_job`. Same as below

Comment (ydye) [reply contrib/api-stress-test/roles/locust/config/templates/stress.py:66]: Done

File: contrib/api-stress-test/roles/openpai/install/defaults/main.yml (hunk 6 above)
Comment (Binyang2014) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: Can we doc about this? If we want to use latest code to do stress test, want steps need to follow.
I think we may need to:
1. checkout master branch
2. apply code change. (Maybe make it as a patch is more user-friendly, but manual modify is OK)
3. push the change to 'stress-test' branch or other branch
4. Change `openpai_branch_name`, run script

Comment (ydye) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: Master branch always changes. I think apply patch to master branch is not good.

Comment (ydye) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: Stress-test can only be a case or example for user who wanna try this stress test environment. And if user wanna use it to test his own code. I still suggest to change the code based on his branch. And I think the commit example is enough for the developer.

Comment (ydye) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: If the user just want to try this environment. Driectly running ansible-playbooks is ok.

Comment (Binyang2014) [reply contrib/api-stress-test/roles/openpai/install/defaults/main.yml:10]: Got it, thanks

**Pull request merged**